### ISSUE TITLE
Consistent CLI executions

### ISSUE THREAD

**Issue by sageprice** [GoogleCloudPlatform/cloud-code-intellij#2598]
Cloud Code has dependencies on the tools `skaffold` and `kubectl`. Internally, we have a common means of setting up the processes to run commands via `CommandExecutor`, but the way that we invoke these dependencies differ.

We've previously discussed (though I can't find the issue) working towards more consistent internal execution of these CLI tools. Particularly, we should be consistent about how we model execution results, and what should happen during an execution.

Proposal:
- Executors should always check if the CLI dependency is available before attempting execution (`SkaffoldExecutorService` currently does not do this)
- General execution methods should return an instance of some sealed class capturing success, failure, or dependency unavailable (both executors currently just throw on failure)
- Executors should support synchronous and asynchronous calls (already done, though naming differs slightly)

Thoughts? Any sore points?

**Comment by sageprice** [comment]
The big question about this is what to do when there is an error executing a process which should display to users, and where these should be surfaced (since in some cases like the run config editor we have to throw exceptions). Curious if anyone has thoughts on where that boundary should lie.

**Comment by etanshaul** [comment]
agree with everything listed in the PR description.

> The big question about this is what to do when there is an error executing a process which should display to users, and where these should be surfaced (since in some cases like the run config editor we have to throw exceptions). Curious if anyone has thoughts on where that boundary should lie.

Doesn't this depend on the context? In the run config, like you said, we need to display the error there. But we also need to display the error in the deployment console (when Skaffold runs) since I think it's possible still to execute a run config that is in an error'ed state. Maybe I missed the point of the question?

**Comment by sageprice** [comment]
I suppose I'm trying to figure out if it is worth returning success/failures or if this is a case where we always want to surface errors. And I think you're right - the best answer is "depends on context."

Was talking to @ZariaHoward about this w.r.t. `KubectlApi`. For that case, the API is just a wrapper that's formatting output. Due to that, it should always fail when a process fails, so returning a success/fail from the executor just means that we've pushed the error throwing logic up into the API. If that winds up being the case everywhere, then there isn't much cause for:

> General execution methods should always return an instance of some sealed class capturing success, failure, or dependency unavailable (both executors currently just throw on failure)

**Comment by ZariaHoward** [comment]
> agree with everything listed in the PR description.
> 
> > The big question about this is what to do when there is an error executing a process which should display to users, and where these should be surfaced (since in some cases like the run config editor we have to throw exceptions). Curious if anyone has thoughts on where that boundary should lie.
> 
> Doesn't this depend on the context? In the run config, like you said, we need to display the error there. But we also need to display the error in the deployment console (when Skaffold runs) since I think it's possible still to execute a run config that is in an error'ed state. Maybe I missed the point of the question?

We should definitely be displaying these errors when appropriate, but I think the consumer of Command Executor should be handling that. The Command Executor should really be limiting its scope to handling the command, not making decisions on what to do with the output. At least that's the approach me and @ivanporty took when designing `KubectlApi`

**Comment by etanshaul** [comment]
I agree with how @ZariaHoward put it that the executors only deal with executing the command and that the consumers decide what do with output (or failure to get the output).

Regarding KubctlApi throwing exceptions: I think perhaps we should revisit this. For example, one thing KubectlApi does is fetch pods (I think). We don’t necessarily want to throw an uncaught exception (consequently surfacing an IDE error) when this happens unless it is truest an exceptional condition: consider a failure to fetch pods when expanding a node in the explorer tree. A much better experience here would be to show an error to the user (somehow), allowing them to refresh and try again.

**Comment by ZariaHoward** [comment]
@etanshaul to your point about the pods. It's not that we aren't catching the error. It's just that we don't catch it in the Kubectl API, we catch it from the UI and decide from the UI code what error to show.

**Comment by etanshaul** [comment]
Oh right. Then I think see @sageprice point then about pushing the error throwing logic up to the API if we return sealed classes from the executors.

Specifically, we’d have to unravel the success / failure then rewrap in a sealed class with the expected return type of the API to be then handled by the caller. Was this your point? In the case of having to propagate errors through multiple layers of the stack, it seems to me like exceptions can be cleaner. But worth discussing.

**Comment by sageprice** [comment]
Minor follow up - I did some testing with [Kotlin's Result monad](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/index.html) ([after enabling it](https://github.com/Kotlin/KEEP/blob/master/proposals/stdlib/result.md#limitations)) to see if it would provide a good way to simplify the error handling. It works well when actually running the plugin, but unfortunately there appears to be an [open issue with Kotlin inline classes](https://youtrack.jetbrains.com/issue/KT-30223) which makes testing via Mockk/Mockito a bit of a non-starter. Definitely worth following up on once the issue is fixed, but it seems like such a solution is not viable at the moment.

I'd argue that once that is available, it should be our preferred approach for dealing with these kinds of behavior.
